pkg/controller/build/imagebuilder/cleaner.go

```diff
@@ -0,0 +1,155 @@
+
+	if err != nil {
+		return fmt.Errorf("could not list configmaps: %w", err)
+	}
+
+	for _, configmap := range configmaps.Items {
+		if err := c.deleteConfigMap(ctx, configmap.Name, mosbName); err != nil {
+			return fmt.Errorf("could not delete ephemeral configmap %s: %w", configmap.Name, err)
+		}
+	}
+
+	secrets, err := c.kubeclient.CoreV1().Secrets(ctrlcommon.MCONamespace).List(ctx, metav1.ListOptions{
+		LabelSelector: selector.String(),
+	})
+
+	if err != nil {
+		return fmt.Errorf("could not list secrets: %w", err)
+	}
+
+	for _, secret := range secrets.Items {
+		if err := c.deleteSecret(ctx, secret.Name, mosbName); err != nil {
+			return fmt.Errorf("could not delete ephemeral configmap %s: %w", secret.Name, err)
+		}
+	}
+
+	return nil
+}
+
+// Deletes a given ConfigMap and tolerates that it was not found so that if
+// this is called more than once, it will not error.
+func (c *cleanerImpl) deleteConfigMap(ctx context.Context, cmName, mosbName string) error {
+	err := c.kubeclient.CoreV1().ConfigMaps(ctrlcommon.MCONamespace).Delete(ctx, cmName, metav1.DeleteOptions{})
+	if err == nil {
+		klog.Infof("Deleted ephemeral ConfigMap %q for build %q", cmName, mosbName)
+		return nil
+	}
+
+	if k8serrors.IsNotFound(err) {
+		return nil
+	}
+
+	return err
+}
+
+// Deletes a given Secret and tolerates that it was not found so that if
+// this is called more than once, it will not error.
+func (c *cleanerImpl) deleteSecret(ctx context.Context, secretName, mosbName string) error {
+	err := c.kubeclient.CoreV1().Secrets(ctrlcommon.MCONamespace).Delete(ctx, secretName, metav1.DeleteOptions{})
+	if err == nil {
+		klog.Infof("Deleted ephemeral secret %q for build %q", secretName, mosbName)
+		return nil
+	}
+
+	if k8serrors.IsNotFound(err) {
+		return nil
+	}
+
+	return err
+}
+
+// Instantiates the selector to use for looking up ConfigMaps and Secrets to
+// delete. It will either use the MachineOSBuild / MachineOSConfig objects or
+// it will get those fields from the Builder object.
+func (c *cleanerImpl) getSelectorForDeletion() (labels.Selector, error) {
+	if c.mosb != nil {
+		// This function can tolerate having MachineOSConfig be nil.
+		return utils.EphemeralBuildObjectSelectorForSpecificBuild(c.mosb, c.mosc)
+	}
+
+	return ephemeralBuildObjectSelectorForBuilder(c.builder)
+}
+
+// Resolve circular import issues.
+func ephemeralBuildObjectSelectorForBuilder(builder buildrequest.Builder) (labels.Selector, error) {
+	mcpName, err := builder.MachineConfigPool()
+	if err != nil {
+		return nil, fmt.Errorf("could not get selector, missing MachineConfigPool: %w", err)
+	}
+
+	renderedMC, err := builder.RenderedMachineConfig()
+	if err != nil {
+		return nil, fmt.Errorf("could not get selector, missing rendered MachineConfig: %w", err)
+	}
+
+	moscName, err := builder.MachineOSConfig()
+	if err != nil {
+		return nil, fmt.Errorf("could not get selector, missing MachineOSConfig: %w", err)
+	}
+
+	return labels.SelectorFromSet(map[string]string{
```

Comment:
non-blocking praise:
Cool way to use selectors here! :+1:

Reply:
Thanks! I was trying to do more with selectors because previously, we were computing the names of all of the various objects which made things cumbersome. There are some really nice helpers in the `labels` package that can be used for this purpose.